### PR TITLE
D8/9 Replace CiviCRM billing fields with Webform fields.

### PR DESCRIPTION
**Pull request by jitendrapurohit**
Overview
----------------------------------------
Replace billing fields with webform fields.

Before
----------------------------------------
Billing fields are loaded using ajax call.

After
----------------------------------------

Billing fields are webform fields. Can be modified to use additional webform benefits, eg layout, conditionals, etc.

image

### CODE REVIEW

File: src/AdminForm.php [hunk 1]
```diff
@@ -1106,6 +1106,40 @@ private function buildContributionTab() {
...
+    $n = wf_crm_aval($this->data, "billing:number_number_of_billing", 0);
+    $this->form['contribution']['sets']["billing_1_number_of_billing"] = [
+      '#type' => 'select',
+      '#title' => t('Enable Billing Address?'),
+      '#default_value' => $n,
+      '#options' => [t('No'), t('Yes')],
+      '#prefix' => '<div class="number-of">',
+      '#suffix' => '</div>',
+      '#description' => t('Enable this section if you want billing address to be sent to the payment processor.'),
+    ];
```
Comment (KarinG) [inline src/AdminForm.php:1126]: Yes to require all billing fields as Civi is expecting them (even if the payment processor couldn't care less). Except for Billing Middle Name - that's not required.

File: src/AdminForm.php [hunk 2]
```diff
@@ -1660,7 +1694,7 @@ public function rebuildData() {
           $this->data[$ent][$c][$key] = $val;
         }
         // Standalone entities keep their own count independent of contacts
-        elseif ($ent == 'grant' || $ent == 'activity' || $ent == 'case' || $ent == 'lineitem' || $ent == 'receipt') {
+        elseif (in_array($ent, ['grant', 'activity', 'case', 'lineitem', 'receipt', 'billing'])) {
           $this->data[$ent]["number_$key"] = $val;
         }
```
Comment (KarinG) [inline src/AdminForm.php:1706]: Nice little clean up!

File: src/AdminForm.php [hunk 3]
```diff
@@ -2194,6 +2231,9 @@ public static function addFieldset($c, &$field, &$enabled, $settings, $ent = 'co
         // 'weight' => $c,
       ];
       $sets = $utils->wf_crm_get_fields('sets');
+      if ($type == 'billing_1_number_of_billing') {
+        $new_set['parent'] = 'contribution_pagebreak';
+      }
       if (isset($isCustom, $customGroupKey)) {
```
Comment (KarinG) [inline src/AdminForm.php:2253]: Yes let's make sure they land on the same page as the Credit Card fields.

File: src/Controller/AjaxController.php [hunk 4]
```diff
@@ -52,30 +52,31 @@ public function handle($key, $input = '') {
...
-            $data = $utils->wf_crm_apivalues('address', 'getoptions', $params);
-        }
-        // @todo use Drupal's cacheable response?
-        return new JsonResponse($data);
+    // @todo use Drupal's cacheable response?
+    return new JsonResponse($data);
+  }
+
+  protected function county($input, $isBilling = FALSE) {
+    $data = [];
```
Comment (KarinG) [inline src/Controller/AjaxController.php:70]: `county`? Does `county` make it into CiviCRM native Billing Fields?

File: src/WebformCivicrmPreProcess.php [hunk 5]
```diff
@@ -404,6 +408,36 @@ private function loadURLEvents($c) {
     }
   }
 
+  /**
+   * Load Billing Address for contact.
+   */
+  private function loadBillingAddress($cid) {
```
Comment (KarinG) [inline src/WebformCivicrmPreProcess.php]: Right - we may have one for an existing / logged in / contact.

File: src/Controller/AjaxController.php (hunk 4 above)
Comment (jitendrapurohit) [reply src/Controller/AjaxController.php:70]: I think this was added to make the function params compatible with [how it is called](https://github.com/colemanw/webform_civicrm/blob/8.x-5.x/src/Controller/AjaxController.php#L46). I agree it looks a bit odd. will fix in the next update. Thanks for pointing 👍

**Comment by KarinG** [comment]
Oh it is all passing now - this is super exciting! I'll block off time for a full round of review and manual testing - hopefully today [if not -> tomorrow].

**Comment by sara-raab** [comment]
I've tested and I can't break it.

**Comment by jitendrapurohit** [comment]
ok - this is all done 👍 . removing wip. 

- Added an upgrade script to add billing fields on existing webforms based on the payment processor enabled on it.
- No billing fields added on the webforms where pp does not expect it to load. Eg pay later webforms.
- If a webform has multiple payment processors with one of them expecting billing fields, the upgrade step adds billing section to these webforms.

**Comment by KarinG** [comment]
This is SO exciting. And it's working! Next I want to run some tests on live forms. I know... but better with my own sites :-) Will try do that tomorrow.

**Comment by KarinG** [comment]
Ok I'm going to run one more test: applying this PR to an existing Webform w/ Contribution/Credit Card payment -> documenting my testing:

```
root@28b67a5d2d33:/var/www/drupal/web/sites/default/modules/webform_civicrm# curl https://patch-diff.githubusercontent.com/raw/colemanw/webform_civicrm/pull/534.diff --output 534.diff
root@28b67a5d2d33:/var/www/drupal/web/sites/default/modules/webform_civicrm# git apply 534.diff
root@28b67a5d2d33:/var/www/drupal/web/sites/default/modules/webform_civicrm# drush updatedb
```

```
Do you wish to run the specified pending updates? (yes/no) [yes]:
>  [notice] Update started: webform_civicrm_update_8003
>  [notice] Update completed: webform_civicrm_update_8003
 [success] Finished performing updates.
```

For good measure:
`root@28b67a5d2d33:/var/www/drupal/web/sites/default/modules/webform_civicrm# sudo -u www-data drush cr
`

I am removing Billing Middle Name from the Webform -> Settings -> CiviCRM -> Contribution tab b/c I now can! 🙂

Testing w/ Live Test Visa -> w/ iATS 1st Pay cryptogram

image

image

Drupal dblog check -> 👍

image

**Comment by KarinG** [comment]
Ok merging this 🎉 Making a note to remind us to check that the tests running on Merge are all ✅ as well.

**Comment by KarinG** [comment]
Noting that this PR initially failed testing b/c there was an issue with Github Actions -> on merge -> tests ran cleanly:

image

**Pull request merged**